Added `wrapNetworkRequest` hook

BrianMcBrayer opened the pull request:
Overview
---------------

This adds a new hook, `wrapNetworkRequest`, which a plugin can use to inspect, log, deny, change, etc. any network requests that go through `httpUtils`.

Changes
---------------

- Add `wrapNetworkRequest` hook
- Add a happy path test to `httpUtils` including all the pieces needed to add additional tests to methods in that module
- Add a specific test for `wrapNetworkRequest` to ensure it's called from `httpUtils.request`. This test verifies also that the correct args are passed to the hook. 

[x] Approach approved
[x] Code written
[x] Tests written & passing
[x] Documentation added

BrianMcBrayer commented on packages/plugin-npm/sources/npmHttpUtils.ts:
```diff
@@ -66,7 +70,12 @@ export async function get(path: string, {configuration, headers, ident, authType
+    const url = path.charAt(0) === `/` ? `${registry}${path}` : path;
+
+    if (typeof rest.reportInfo !== `undefined`)
+      rest.reportInfo(MessageName.UNNAMED, `Fetching ${url}`);
```
If we decide to log from here, I am not sure what `MessageName` to use. Is `UNNAMED` sufficient? Should I create a new one?

BrianMcBrayer replied:
I guess I should also deconstruct `reportInfo` from `rest` instead of getting it off of `rest`. I just wanted to highlight the fundamentals of this change.

Also, there might be other code that calls this where we would also want to pass down `reportInfo` -- that way we could get a comprehensive audit trail in the logs of all the packages we attempted to pull.

This might help in investigations into dependency confusion attacks as well.

arcanis commented:
Some comments:

- Better pass the whole `report` as option rather than just `reportInfo`
- Better use `reportJson` than `reportInfo` - the former is intended for programmatic use, whereas the later is for human consumption. Humans won't care much for the url, especially since it will be almost the same from one message to the other.
- I'd think a possibly better approach (or, at least, complementary) would be to create a `wrapNetworkRequest` hook, similar to what we already do with [`wrapScriptExecution`](https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-core/sources/scriptUtils.ts#L419-L429). It'd also be handy in general to block requests, monitor their timing, mock the network (would be useful to eventually make Yarn work in-browser), etc.

BrianMcBrayer commented:
Awesome. I'll try to implement this today.

Thanks for the feedback.

BrianMcBrayer commented:
> Some comments:
> 
> * Better pass the whole `report` as option rather than just `reportInfo`
> * Better use `reportJson` than `reportInfo` - the former is intended for programmatic use, whereas the later is for human consumption. Humans won't care much for the url, especially since it will be almost the same from one message to the other.
> * I'd think a possibly better approach (or, at least, complementary) would be to create a `wrapNetworkRequest` hook, similar to what we already do with [`wrapScriptExecution`](https://github.com/yarnpkg/berry/blob/master/packages/yarnpkg-core/sources/scriptUtils.ts#L419-L429). It'd also be handy in general to block requests, monitor their timing, mock the network (would be useful to eventually make Yarn work in-browser), etc.

@arcanis I really like your `wrapScriptExecution` hook idea. That is the route that I took here.

Now I need advice. The `Report` object is not passed into `httpUtils` currently. Would the correct approach be to add it to the params passed to each public method (`request`, `get`, `put`, `del`, `post`, etc.) there and then plumb that up? That is not that big of an effort I don't think. Or... is there a better way to get the reporter than to go plumbing it down? Thoughts?

After we get that worked out, I'll add some tests and also write the plugin that I want in my own repo (unless we want that included in Yarn).

BrianMcBrayer commented on packages/yarnpkg-core/sources/Plugin.ts:
```diff
@@ -9,6 +11,7 @@ import {MessageName}
 import {Project, InstallOptions}                                                              from './Project';
 import {Resolver, ResolveOptions}                                                             from './Resolver';
 import {Workspace}                                                                            from './Workspace';
+import {Body, Method}                                                                         from './httpUtils';
```
Should I import these from `httpUtils`, or should I hoist them from `httpUtils` into types.ts?

BrianMcBrayer replied:
Still unsure of what yall would prefer here.

arcanis replied:
They are fine in `httpUtils`; however by convention we star-import all `*Utils.ts` files:

```suggestion
import * as httpUtils                                                                         from './httpUtils';
```

BrianMcBrayer replied:
Done

BrianMcBrayer commented on packages/yarnpkg-core/sources/httpUtils.ts:
```diff
@@ -172,6 +172,52 @@ export type Options = {
+  return response.body;
+}
+
+async function requestImpl(target: string | URL, body: Body, {configuration, headers, jsonRequest, jsonResponse, method = Method.GET}: Omit<Options, 'customErrorMessage'>) {
```
This is the exact implementation of `request`, except now it's private. All requests now funnel through this.

BrianMcBrayer commented on packages/yarnpkg-core/sources/Plugin.ts:
```diff
@@ -62,7 +65,7 @@ export type Hooks = {
 
   /**
-   * When a script is getting executed. You must call the executor, or the
+   * Called when a script is getting executed. You must call the executor, or the
```
Changed this to match the other descriptions

merceyz commented on packages/yarnpkg-core/sources/Plugin.ts:
```diff
@@ -73,6 +76,14 @@ export type Hooks = {
+   * Called when a network request is being made.
+   */
+  wrapNetworkRequest?: (
+    networkRequest: () => Promise<Response<any>>,
```
I don't think we should make our public interface depend on `got`

BrianMcBrayer replied:
Yeah, that's a good point. Hmm. We could make our own type that has the same shape? Or something else?

BrianMcBrayer replied:
Done. I think.

Since `request` could really return just about anything, I changed the plugin to return `Promise<any>` rather than creating a matching type for `Response`, which is an involved type. I think that should be sufficient? Please let me know.

BrianMcBrayer commented:
@merceyz or @arcanis would one of you mind reviewing this when you can so that I can either make adjustments or add a couple unit/integration tests?

Also... once this is merged, I might take a stab at documenting the existing plugins if my work schedule permits.

BrianMcBrayer commented on .yarn/versions/d075ed96.yml:
```diff
@@ -0,0 +1,4 @@
+undecided:
```
I have no clue what to do here. I'd say maybe that `@yarnpkg/core` should get a minor bump because we added the hook? And the others could be `patch`?

arcanis replied:
Yep for the core 👍 On top of it, `@yarnpkg/cli` should also be minor because it's the source of truth for the bundled CLI we release.

Other packages can be set to "declined", since they don't mandate this new feature.

BrianMcBrayer replied:
Done

arcanis commented on packages/yarnpkg-core/sources/Plugin.ts:
```diff
@@ -73,6 +75,14 @@ export type Hooks = {
+   */
+  wrapNetworkRequest?: (
+    networkRequest: () => Promise<any>,
+    extra: {target: string | URL, body: Body, configuration: Configuration, headers: { [headerName: string]: string } | undefined, jsonRequest: boolean | undefined, jsonResponse: boolean | undefined, method: Method | undefined}
+  ) => Promise<() => Promise<any>>;
```
I'd prefer if this had its own type. Somewhere at the beginning of the file you can do something like:

```ts
export type WrapNetworkRequestInfo = httpUtils.Options & {
  target: string | URL;
  body: Body;
};
```

Then reuse this type in the parameter (and don't forget to export this type from `index.ts`).

BrianMcBrayer replied:
Done

BrianMcBrayer commented:
I'll write some tests over the next couple of days and then flip this PR from a draft to ready for review. Thank you all for all the suggestions.

BrianMcBrayer commented:
Hey folks. I added two tests (one happy path test in general to make sure I didn't break anything -- didn't find any existing tests on `httpUtils`; one test for the functionality that we added here).

I _think_ this is ready for a real review. Please let me know what I should change.

BrianMcBrayer commented:
These integration tests seem flaky... The past two commits have had different failures, and the last commit was just tweaking a code comment.

arcanis commented:
Looks good to me, thanks @BrianMcBrayer !
